### PR TITLE
Implement MIX invitation

### PR DESCRIPTION
**Pull request by melvo**
The invitation is specified by XEP-0407: Mediated Information eXchange (MIX): Miscellaneous Capabilities.

### CODE REVIEW

File: src/base/QXmppMessage.cpp [hunk 1]
```diff
@@ -1073,6 +1077,36 @@ void QXmppMessage::setSpoilerHint(const QString &spoilerHint)
         d->isSpoiler = true;
 }
 
+///
+/// Determines whether this message is a \xep{0369}: Mediated Information eXchange (MIX) invitation message as defined by \xep{0407}: Mediated Information eXchange (MIX): Miscellaneous Capabilities.
```
Comment (lnjX) [inline src/base/QXmppMessage.cpp:1081]: 80 char line limit please :)

File: src/base/QXmppMessage.cpp [hunk 2]
```diff
@@ -167,6 +168,9 @@ class QXmppMessagePrivate : public QSharedData
     bool isSpoiler;
     QString spoilerHint;
 
+    // XEP-0407: Mediated Information eXchange (MIX): Miscellaneous Capabilities
+    QXmppMixInvitation mixInvitation;
```
Comment (lnjX) [inline src/base/QXmppMessage.cpp:172]: We can use `std::optional` for that, that's more expressive.

File: src/base/QXmppMessage.cpp [hunk 3]
```diff
@@ -1073,6 +1077,36 @@ void QXmppMessage::setSpoilerHint(const QString &spoilerHint)
         d->isSpoiler = true;
 }
 
+///
+/// Determines whether this message is a \xep{0369}: Mediated Information eXchange (MIX) invitation message as defined by \xep{0407}: Mediated Information eXchange (MIX): Miscellaneous Capabilities.
+///
+/// \since QXmpp 1.5
```
Comment (lnjX) [inline src/base/QXmppMessage.cpp:1083]: If you're targeting `1.4` (which is going to be the next release) with this, then please use that everywhere.

File: src/base/QXmppMessage.cpp [hunk 4]
```diff
@@ -1073,6 +1077,36 @@ void QXmppMessage::setSpoilerHint(const QString &spoilerHint)
...
+///
+bool QXmppMessage::isMixInvitation() const
+{
+    return !d->mixInvitation.isNull();
+}
```
Comment (lnjX) [inline src/base/QXmppMessage.cpp:1088]: With the optional this isn't needed anymore.

File: src/base/QXmppMessage.cpp [hunk 5]
```diff
@@ -1327,6 +1361,11 @@ void QXmppMessage::toXml(QXmlStreamWriter *xmlWriter) const
         xmlWriter->writeEndElement();
     }
 
+    // XEP-0407: Mediated Information eXchange (MIX): Miscellaneous Capabilities
+    if (!d->mixInvitation.isNull()) {
+        d->mixInvitation.toElement().toXml(xmlWriter);
```
Comment (lnjX) [inline src/base/QXmppMessage.cpp:1366]: Please don't use QXmppElement when it's possible without it.

File: src/base/QXmppMixInvitation.cpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+void QXmppMixInvitation::parse(const QXmppElement &element)
+{
+    d->inviterJid = element.firstChildElement(QStringLiteral("inviter")).value();
+    d->inviteeJid = element.firstChildElement(QStringLiteral("invitee")).value();
+    d->channelJid = element.firstChildElement(QStringLiteral("channel")).value();
+    d->token = element.firstChildElement(QStringLiteral("token")).value();
+}
+
+QXmppElement QXmppMixInvitation::toElement() const
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.cpp:149]: This should be done with the QXmlStreamWriter (and parsing with QDomElements)

**Review by lnjX — CHANGES_REQUESTED** [review]

File: tests/qxmppmixinvitation/tst_qxmppmixinvitation.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            "<invitee>cat@shakespeare.example</invitee>"
+            "<channel>coven@mix.shakespeare.example</channel>"
+            "<token>ABCDEF</token>"
+        "</invitation>"
+    );
+
+    QXmppMixInvitation invitation;
+    parsePacket(invitation, xml);
+
+    QCOMPARE(invitation.inviterJid(), "hag66@shakespeare.example");
```
Comment (lnjX) [inline tests/qxmppmixinvitation/tst_qxmppmixinvitation.cpp:53]: You need to use `QCOMPARE(invitation.inviterJid(), QString("hag66@shakespeare.example"))` or `QCOMPARE(invitation.inviterJid(), QStringLiteral("hag66@shakespeare.example"))`. This is causing the compile error on Qt 5.9.

File: src/base/QXmppMixInvitation.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    QString inviteeJid() const;
+    void setInviteeJid(const QString &inviteeJid);
+
+    QString channelJid() const;
+    void setChannelJid(const QString &channelJid);
+
+    QString token() const;
+    void setToken(const QString &token);
+
+    bool isNull() const;
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.h:61]: This can be removed.

File: src/base/QXmppMixInvitation.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+#define QXMPPMIXINVITATION_H
+
+#include <QSharedDataPointer>
+
+#include "QXmppElement.h"
+
+class QXmppMixInvitationPrivate;
+
+///
+/// \brief The QXmppMixInvitation class is used to invite a user to a \xep{0369}: Mediated Information eXchange (MIX) channel as defined by \xep{0407}: Mediated Information eXchange (MIX): Miscellaneous Capabilities.
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.h:34]: 80 char limit

File: src/base/QXmppMixInvitation.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+///
+/// \ingroup Stanzas
+///
+/// \since QXmpp 1.5
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.h:38]: 1.4

File: src/base/QXmppMixInvitation.cpp [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+///
+/// Determines whether the given DOM element is a MIX invitation.
+///
+/// \param element DOM element being checked
+///
+/// \return true if element is a MIX invitation, otherwise false
+///
+bool QXmppMixInvitation::isMixInvitation(const QDomElement &element)
+{
+    return element.tagName() == QStringLiteral("invitation") && (element.namespaceURI().isEmpty() || element.namespaceURI() == ns_mix_misc);
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.cpp:175]: In which case should we accept an empty namespace? As namespaces are inherited, the namespace is never going to be empty in an XMPP stream.

File: src/base/QXmppMixInvitation.cpp [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+///
+/// \return the generated token used for authentication
+///
+QString QXmppMixInvitation::token() const
+{
+    return d->token;
+}
+
+///
+/// Sets the token which is generated by the server and used by the invitee for authentication.
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.cpp:123]: 80 char limit

File: src/base/QXmppMessage.h [hunk 13]
```diff
@@ -243,6 +245,10 @@ class QXMPP_EXPORT QXmppMessage : public QXmppStanza
     QString spoilerHint() const;
     void setSpoilerHint(const QString &);
 
+    // XEP-0407: Mediated Information eXchange (MIX): Miscellaneous Capabilities
+    std::optional<QXmppMixInvitation> mixInvitation() const;
+    void setMixInvitation(const QXmppMixInvitation &mixInvitation);
```
Comment (lnjX) [inline src/base/QXmppMessage.h:250]: This should also be an `std::optional`, so you can remove the invitation with `std::nullopt`.

File: src/base/QXmppMessage.cpp [hunk 14]
```diff
@@ -1327,6 +1355,11 @@ void QXmppMessage::toXml(QXmlStreamWriter *xmlWriter) const
         xmlWriter->writeEndElement();
     }
 
+    // XEP-0407: Mediated Information eXchange (MIX): Miscellaneous Capabilities
+    if (!d->mixInvitation) {
```
Comment (lnjX) [inline src/base/QXmppMessage.cpp:1359]: Isn't the `!` wrong here?

File: src/base/QXmppMessage.h [hunk 15]
```diff
@@ -26,6 +26,8 @@
 #ifndef QXMPPMESSAGE_H
 #define QXMPPMESSAGE_H
 
+#include "QXmppMixInvitation.h"
```
Comment (lnjX) [inline src/base/QXmppMessage.h:29]: this can be forward declared

File: src/base/QXmppMixInvitation.cpp [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+
+void QXmppMixInvitation::parse(const QDomElement &element)
+{
+    d->inviterJid = element.firstChildElement(QStringLiteral("inviter")).text();
+    d->inviteeJid = element.firstChildElement(QStringLiteral("invitee")).text();
+    d->channelJid = element.firstChildElement(QStringLiteral("channel")).text();
+    d->token = element.firstChildElement(QStringLiteral("token")).text();
+}
+
+void QXmppMixInvitation::toXml(QXmlStreamWriter* writer) const
```
Comment (lnjX) [inline src/base/QXmppMixInvitation.cpp:150]:
```suggestion
void QXmppMixInvitation::toXml(QXmlStreamWriter *writer) const
```

**Review by lnjX — CHANGES_REQUESTED** [review]
And the tests are still failing.

File: src/base/QXmppMessage.h (hunk 15 above)
Comment (melvo) [reply src/base/QXmppMessage.h]: I tried it but it does not build when I use `class QXmppMixInvitation;`.

Comment (lnjX) [reply src/base/QXmppMessage.h:29]: I tested it and it works for me, what's the error message?

Comment (melvo) [reply src/base/QXmppMessage.h]: I had to add `#include "QXmppMixInvitation.h"` to `tests/qxmppmessage/tst_qxmppmessage.cpp`. bbf612a6b294bc73dcf9a741b9ce78b4c63f6b3a should fix it.

File: tests/qxmppmixinvitation/tst_qxmppmixinvitation.cpp [hunk 17]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2008-2020 The QXmpp developers
```
Comment (lnjX) [inline tests/qxmppmixinvitation/tst_qxmppmixinvitation.cpp:2]:
```suggestion
 * Copyright (C) 2008-2021 The QXmpp developers
```
(I just updated the copyright headers)

**Review by lnjX — CHANGES_REQUESTED** [review]
```
The following tests FAILED:
         16 - tst_qxmppmixinvitation (Failed)
```

File: tests/qxmppmixinvitation/tst_qxmppmixinvitation.cpp [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        "<invitation>"
+            "<inviter>hag66@shakespeare.example</inviter>"
+            "<invitee>cat@shakespeare.example</invitee>"
+            "<channel>coven@mix.shakespeare.example</channel>"
+            "<token>ABCDEF</token>"
+        "</invitation>"
+    );
+    QCOMPARE(doc.setContent(correctInvitationWithoutNamespace, true), true);
+    element = doc.documentElement();
+    QVERIFY(QXmppMixInvitation::isMixInvitation(element));
```
Comment (lnjX) [inline tests/qxmppmixinvitation/tst_qxmppmixinvitation.cpp:99]: This is outdated / should be false now. After that the test should also pass again.

**Comment by lnjX** [comment]
You disabled push access to your branch. Merge https://github.com/olomono/qxmpp/pull/1 please.

File: src/base/QXmppMessage.h [hunk 19]
```diff
@@ -26,13 +26,18 @@
...
-// Required for source compatibility
 #include "QXmppStanza.h"
 
+// Required for source compatibility
 #include <QDateTime>
 
+namespace std {
+template<typename T>
+class optional;
+}
```
Comment (lnjX) [inline src/base/QXmppMessage.h:37]: The macOS test is failing now, we should probably just include the optional instead of forward-declaring it here.

**Review by lnjX — APPROVED** [review]
Everything correct now! :)

**Pull request merged**